---
title: "data_long_generation"
author: "wenyu"
date: "May 14, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r}
library(tidyverse)
library(data.table)
library(openxlsx)
options(stringsAsFactors = FALSE)
setwd("C:/Users/wenyu/Documents/functional genomics/manuscript3_ebiomedicine/data_long_generation")
memory.limit(size = 30000)
```

combine crispr based ess score from different studies into a big df
the crispr score for cell lines that has been screened by RNAi in demeter paper were colleced from previous 5 studies.


```{r}
cas9_1 <- read.xlsx("data crispr.xlsx","list1")
cas9_2 <- read.xlsx("data crispr.xlsx","list2")
cas9_3 <- read.xlsx("data crispr.xlsx","list3")
cas9_4 <- read.xlsx("data crispr.xlsx","list4")
cas9_5 <- read.xlsx("data crispr.xlsx","list5")
# name <- colnames(cas9_5)
# cas9_5 <- cas9_5[,2:35]
# colnames(cas9_5) <- c("gene", name[3:35])

##since list1 and 2 are using Bagel, in which the larger the ess score the more essential the gene is.

# to make them comparable with other studies we take a negative of all the values in this two data source

summary(cas9_1)
cas9_1[,2:3] <- -cas9_1[,2:3]
summary(cas9_1)
summary(cas9_2)
cas9_2[,2:3] <- -cas9_2[,2:3]



cas9 <- cas9_1 %>% 
  full_join(cas9_2) %>% 
  full_join(cas9_3) %>% 
  full_join(cas9_4) %>% 
  full_join(cas9_5) 




summary(cas9)  
apply(cas9, 2, min,na.rm=T)

#note that DLD1 and HCT116 have extremely large ess scores 
```




#shRNA data from two different screen has been averaged into gene level score 
```{r}
rnai1 <- fread("Achilles_v2.4.6.rnai.gct")
rnai2 <- fread("achilles-v2-19-2-mapped-rnai_v1-data.gct")
rnai1 <- rnai1[,2:218]
rnai2 <- rnai2[,2:287]
rnai1 <- rnai1 %>% group_by(Description) %>% summarise_all(mean,na.rm = TRUE)
rnai2 <- rnai2 %>% group_by(Description) %>% summarise_all(mean,na.rm = TRUE)
shRNA <- full_join(rnai1,rnai2,"Description")
# write.csv(rnai,"rnai.csv",row.names = FALSE)
# shRNA <- fread("C:/Users/wenyu/Documents/functional genomics/data collection/rnai_un/rnai.csv")
cell <- colnames(cas9)[2:43]
shRNA <- tbl_df(shRNA) %>% 
  select(one_of(c("Description", cell))) %>% 
  arrange(Description) 
colnames(shRNA)[1] <- "gene"
```




```{r}
#there is no ccle data for DLD1
cell_mole <- setdiff(cell, c("DLD1_LARGE_INTESTINE"))
rnaseq <- fread("CCLE_RNAseq_081117.rpkm.gct")
rnaseq <- rnaseq %>% 
  select(one_of(c("Description",cell_mole )))%>% 
  mutate(DLD1_LARGE_INTESTINE = NA) %>% 
  arrange(Description)
colnames(rnaseq)[1] <- "gene"
```

```{r}

rnaarray <- fread("CCLE_Expression_Entrez_2012-09-29.gct")
# there are two column for NCIH292_LUNG
check <- colnames(rnaarray)
check1 <- which(check %in% "NCIH292_LUNG")
check[c(965,993)]
cor(rnaarray[,c(965,993)]) #highly correlated, will be considered as technical duplicate, mean value of the two column is used
NCIH292_LUNG <- cbind(rnaarray[,965],rnaarray[,993])
NCIH292_LUNG <- apply(NCIH292_LUNG,1,mean)
rnaarray <- rnaarray[,c(-965,-993)]

rnaarray <- rnaarray %>% 
  mutate(NCIH292_LUNG = NCIH292_LUNG) %>% 
  select(one_of(c("Description",cell_mole )))%>% 
  mutate(DLD1_LARGE_INTESTINE = NA) %>% 
  arrange(Description)
colnames(rnaarray)[1] <- "gene"
```

```{r}
cn <- fread("CCLE_copynumber_byGene_2013-12-03.txt")
cn <- cn %>% 
  select(one_of(c("SYMBOL",cell_mole )))%>% 
  mutate(DLD1_LARGE_INTESTINE = NA) %>% 
  arrange(SYMBOL)
colnames(cn)[1] <- "gene"
```

```{r}
mutation <- fread("ccle2maf_081117.txt")

#original mutation data is no in rectangular format but rather a long format of record.
#it was preprocessed into this count matrix
mut <- fread("mutation matrix.txt")
mut <-  mut %>% 
  select(one_of(c("gene",cell_mole )))%>% 
  mutate(DLD1_LARGE_INTESTINE = NA) %>% 
  arrange(gene)
dist(mut)
```


```{r}
demeter <- fread("https://depmap.org/portal/download/api/download/external?file_name=15%2FExpandedGeneZSolsCleaned.csv")
demeter <- tbl_df(demeter) %>% 
  select(one_of(c("V1", cell))) %>% 
  arrange(V1) 
colnames(demeter)[1] <- "gene"
```


#collect ceres and turn into wide matrix
```{r}
#first get ceres scores from gecko and ceres_wang that is avaliable in ceres paper
#then try to search for the rest in the ceres_563
gecko_33 <- read.csv("https://media.nature.com/original/nature-assets/ng/journal/v49/n12/extref/ng.3984-S6.csv")
ceres_wang <- read.csv("https://media.nature.com/original/nature-assets/ng/journal/v49/n12/extref/ng.3984-S7.csv")
dim(gecko_33)
colnames(gecko_33)[1] <-  colnames(ceres_wang)[1] <- "gene"

idx <- which(colnames(gecko_33) %in%  cell )
ceres_df1 <- gecko_33[,c(1,idx)]

idx <- which(colnames(ceres_wang) %in%  cell )
ceres_df2 <- ceres_wang[,c(1,idx)]

#for 42 cells not in the gecko_33 and ceres_wang, try to find them from the ceres_563
idx <- which(!(cell  %in%  c(colnames(ceres_wang),colnames(gecko_33))))
ceres_563 <- fread("https://ndownloader.figshare.com/files/15023465")
cellmap <- fread("https://ndownloader.figshare.com/files/15023525")
colnames(ceres_563)[1] <- "cell"

ceres_563$cell <- cellmap$CCLE_name[match(ceres_563$cell,cellmap$DepMap_ID)]
add_cell <- cell[which(cell[idx] %in% ceres_563$cell)] 
ceres_df3 <- ceres_563 %>%  filter(cell %in%  add_cell )
ASPC1_PANCREAS= unlist(ceres_df3[1,])[-1]
names(ASPC1_PANCREAS) = NULL
DLD1_LARGE_INTESTINE= unlist(ceres_df3[2,])[-1]
names(DLD1_LARGE_INTESTINE) = NULL
gene= colnames(ceres_df3)[-1]
head(gene) #there are numbers in the gene name and should be removed
gene <-  str_split(gene," " , n = 2, T)[,1]
head(gene) 
ceres_df3 <- data.frame(gene=gene, 
                        DLD1_LARGE_INTESTINE= as.numeric(DLD1_LARGE_INTESTINE),
                        ASPC1_PANCREAS= as.numeric(ASPC1_PANCREAS),
                        stringsAsFactors = F)

##now combine the three ceres_df into one
ceres <- ceres_df1 %>% 
  full_join(ceres_df2) %>% 
  full_join(ceres_df3)
#summary(ceres)
```



#create a function to convert all the table into a long one so that it is easier to combine them with each other
```{r}

##check if there are duplicate in genes
cas9_1$gene[which(duplicated(cas9_1$gene))]
cas9_2$gene[which(duplicated(cas9_2$gene))]
cas9_3$gene[which(duplicated(cas9_3$gene))]
cas9_4$gene[which(duplicated(cas9_4$gene))]
cas9_5$gene[which(duplicated(cas9_5$gene))]
shRNA$gene[which(duplicated(shRNA$gene))]
cn$gene[which(duplicated(cn$gene))]
mut$gene[which(duplicated(mut$gene))]
rnaseq$gene[which(duplicated(rnaseq$gene))]
rnaarray$gene[which(duplicated(rnaarray$gene))]
demeter$gene[which(duplicated(demeter$gene))]
ceres$gene[which(duplicated(ceres$gene))]

#function for compressing the duplicates into one row

compress_for_dupgene <- function(df){
  df1 = df %>% group_by(gene) %>% summarise_all(mean,na.rm= T)
  return(df1)
}


make_wide_to_long <- function(df,value){
  leng <- length(df)
  df1 <- gather(df,key = cell,value = val, 2:leng)
  colnames(df1)[3] <- value
  return(df1)
}


cas9_long <- make_wide_to_long(compress_for_dupgene(cas9),"crispr")
shRNA_long <- make_wide_to_long(compress_for_dupgene(shRNA),"rnai_un")
exp.array_long <- make_wide_to_long(compress_for_dupgene(rnaarray),"array")
exp.seq_long <- make_wide_to_long(compress_for_dupgene(rnaseq),"seq")
cn_long <-  make_wide_to_long(compress_for_dupgene(cn),"cn")
mut_long <- make_wide_to_long(compress_for_dupgene(mut),"mut")
ceres_long <- make_wide_to_long(compress_for_dupgene(ceres),"ceres")
demeter_long <- make_wide_to_long(compress_for_dupgene(demeter),"demeter")
```

#combine all the dataset into data_long
```{r}
data_long <-  cas9_long %>% 
  full_join(shRNA_long) %>% 
  full_join(exp.array_long) %>% 
  full_join(exp.seq_long) %>% 
  full_join(cn_long) %>% 
  full_join(mut_long) %>% 
  full_join(ceres_long) %>% 
  full_join(demeter_long)

#filter out rows that are not genes
genesymbol <- scan('C:/Users/wenyu/Documents/functional genomics/pipeline4/mart_export.txt',sep = ',',what = "character",) ##filter the rows that are not genes
data_long1 <- filter(data_long, gene %in% genesymbol)
dim(data_long1)  

data_long1 <- data_long1 %>% 
  mutate(cell_line= str_split(cell,"_",n = 2,simplify = T)[,1],
         tissue_origin = str_split(cell,"_",n = 2,simplify = T)[,2]
         )

#data_long1[,3:9] <- apply(data_long1[,3:9],2 ,as.numeric )
summary(data_long1)
table(data_long1$cell_line)
table(data_long1$tissue_origin)
hist(data_long1$crispr)
hist(data_long1$rnai_un)
hist(data_long1$seq)
hist(data_long1$array)
hist(data_long1$mut)
hist(data_long1$ceres)
hist(data_long1$demeter)






#decide the transfomation for the dataset
# check what is wrong with crispr dataset, why there are extreme values?




write.csv(data_long1,file = "data_long1_new.csv",row.names = F)
```

check the problem of the crispr dataset
```{r}
hist(cas9_2$DLD1_LARGE_INTESTINE,breaks = 15)
hist(scale(cas9_2$DLD1_LARGE_INTESTINE),breaks = 15)
hist(cas9_2$HCT116_LARGE_INTESTINE,breaks = 15)
hist(scale(cas9_2$HCT116_LARGE_INTESTINE),breaks = 15)
```


```{r}

cell_cas33 <- colnames(cas9_5)[-1]
write(cell_cas33,file = "cell_cas33.txt") 
```

